krop.1
.TH krop 1
.SH NAME
krop \- a tool to crop PDF files
.SH SYNOPSIS
.B krop
.RI [ options ]
.RI [ file ]
.br
.SH DESCRIPTION
This manual page briefly documents the
.B krop
command.
.PP
\fBkrop\fP is a simple graphical tool to crop the pages of PDF files. It is written
in Python and relies on PyQt and PyMuPDF (or a suitable subset of pypdf/pikepdf/
python-poppler-qt) for its functionality.  A unique feature of krop is its
ability to automatically split pages into subpages to fit the limited screen
size of devices such as eReaders. This is particularly useful, if your eReader
does not support convenient scrolling.
.SH OPTIONS
A summary of options is included below.
For a complete description, use the \fB\-\-help\fP option.
.SS "Arguments:"
.TP
.B file
PDF file to open
.SS "Options:"
.TP
.B \-h, \-\-help
show help, including all options, and exit
.TP
.B \-v, \-\-version
show program's version number and exit
.TP
.B \-o OUTPUT, \-\-output OUTPUT
where to save the cropped PDF
.TP
.B \-\-whichpages WHICHPAGES
which pages (e.g. "1\-5" or "1,3\-") to include in cropped PDF (default: all)
.TP
.B \-\-rotate {0,90,180,270}
how much to rotate the cropped pdf clockwise (default: 0)
.TP
.B \-\-optimize {gs,no}
whether to optimize the final PDF using ghostscript (default: previous choice)
.TP
.B \-\-grid GRID
if set to 2x3, for instance, creates a 2x3 grid of selections on initial page; if only one number is specified, the number of columns/rows is determined according to whether the page is landscape or portrait
.TP
.B \-\-initialpage INITIALPAGE
which page to open initially (default: 1)
.TP
.B \-\-selections {all,evenodd,individual}
to which pages should selections apply
.TP
.B \-\-exceptions EXCEPTIONS
pages (e.g. "1\-5" or "1,3\-") which require individual selections
.TP
.B \-\-trim
if specified, will auto trim initial selections
.TP
.B \-\-trim-use {initial,all}
whether to inspect only the initial page or all pages (slow!) when auto trimming (default: previous value)
.TP
.B \-\-trim\-padding TRIM_PADDING
how much padding to include when auto trimming (default: previous value)
.TP
.B \-\-go
output PDF without opening the krop GUI (using the choices supplied on the command line); if used in a script without X server access, you can run krop using xvfb\-run


ChangeLog
krop (0.6.1)  2023-00-00 -- Armin Straub

  * Preserve links within a PDF (thanks to chrthi for doing this).
  * Support the new version of pypdf in addition to its predecessor PyPDF2
  (thanks to Emmanuel Rosa for the idea and initial code).
  + Support pikepdf as an alternative to pypdf (thanks to Hongzhuo Liang for
  the idea and initial code).

krop (0.6.0)  2020-06-09 -- Armin Straub

  + Fixed aspect ratios, like letter size, can be chosen for selections.
  + Selections (or a grid of selections) for the full page can now be created
  using the context menu or by pressing Insert (or Shift+Insert). Also
  introduced the command line option --grid to create a grid of selections on
  the initial page.
  + Auto trimming margins can now inspect all pages.
  + New option for whether to include pages without selections in the output.
  + Added the command line options --optimize (thanks to Ondrej Tichacek for
  suggesting this feature) and --exceptions.
  + Implemented several keyboard shortcuts including Shift+Arrow to move
  current selection and Delete to remove it.
  * Keep track of current selection and highlight it visually.
  * Don't fail on PDFs that are encrypted with an empty password.
  * Remember window geometry and fit in view setting.


krop (0.5.1)  2018-10-27 -- Armin Straub

  * Replace and extend the README file with a MarkDown version (thanks to
  Eduardo Montenegro for doing this).
  + Add a manpage.
  * Fix a Qt5 related bug when selecting filename for saving (thanks to
  Lin-Buo-Ren for reporting this).


krop (0.5.0)  2018-02-11 -- Armin Straub

  + Support PyQt5.
  + Use Ghostscript to optionally optimize the final PDF (thanks to Mathias
  Rav for the idea and code).


krop (0.4.13)  2017-07-22 -- Armin Straub

  + Added the command line option --go, which outputs a PDF using the choices
  from --autotrim, --rotate and --whichpages, without opening the krop GUI
  (thanks to Mirko Weihrauch for the idea).
  + Padding for auto trimming can be specified as in CSS using one to four
  values.  This option is now also available on the command line as
  --autotrim-padding.  (Thanks to Jonah Langlieb for providing sanity checks!)
  * Fix incorrect string handling when opening files from the command line
  using python3 (thanks to Giuliano Costantini for observing and fixing this).


krop (0.4.12)  2017-01-22 -- Armin Straub

  * Use CropBox instead of MediaBox when determining the region to be cropped
  (thanks to Tim Mueller for observing and fixing this).
  * Use a more proper Python package structure (thanks to Xavier Corredor
  Llano for advice).


krop (0.4.11)  2016-08-20 -- Armin Straub

  * Make krop compatible with both python2 and python3.


krop (0.4.10)  2016-05-29 -- Armin Straub

  * When cropping, all PDF page boxes (ArtBox, BleedBox, CropBox, MediaBox,
  TrimBox) are now adjusted.


krop (0.4.9)  2015-10-31 -- Armin Straub

  * Reading PDF files is less strict now.  As a consequence, krop doesn't give
  up cropping some files which resulted in an error before.
  * Fix a possible crash when starting krop.


krop (0.4.8)  2015-07-04 -- Armin Straub

  + More options are accessible via command line arguments.
  * Mouse cursor now changes when over selections and handles.
  * Improved visibility of selections.


krop (0.4.7)  2015-02-15 -- Armin Straub

  + Basic options are accessible as command line arguments (run krop with
  --help to see available options).
  * When pressing ctrl-c in the terminal, krop will shut down (not gracefully,
  though; only intended for when there is trouble).
  * Avoid hitting Python's recursion limit in the pyPdf library when working
  with certain large pdf files (thanks to Victor).


krop (0.4.6)  2014-05-24 -- Armin Straub

  + Added option to specify pages which require individual selections (thanks
  to Roger Krebs for the suggestion and for testing).
  * 'Trim Margins' automatically creates a region for the full page if no
  regions exist.
  * Reorganized options across tabs.
  + Added appdata file (thanks to Richard Hughes).


krop (0.4.5)  2014-02-01 -- Armin Straub

  * Load and write files with non-ascii characters in their filename.


krop (0.4.4)  2013-08-18 -- Armin Straub

  + If available use PyPDF2 in place of pyPdf.
  * Correctly crop PDFs which have orientation other than 'portrait' (thanks
  to Gaetano Pepe for sending sample PDFs and testing).
  * In case of more than one selection per page, cropping is much faster now
  and results in smaller files.
  * Delete old selections when opening a new file.


krop (0.4.3)  2013-08-10 -- Armin Straub

  * Fix a bug that prevented PDF files to be specified on the command line
  (thanks to roctheme for reporting).
  * Instead of no icon at all, use the system icon 'edit-cut' as application
  icon (thanks to roctheme for the suggestion).


krop (0.4.2)  2013-08-04 -- Armin Straub

  + Use PyKDE if available (though this is not really used yet).
  * If default icons are missing, use text instead.


krop (0.4.1)  2013-07-28 -- Armin Straub

  * Initial public release.
  + Several small bug fixes.


krop (0.4.0)  2013-06-16 -- Armin Straub

  * Rewrote cropping code, resulting in much smaller pdf files.


krop (0.3.0)  2011-07-14 -- Armin Straub

  + Added option to specify file name for cropped file.
  + Added option to specify which pages to include into the cropped file.


krop (0.2.0)  2010-10-16 -- Armin Straub

  + Added option to automatically trim selection (right-click a selection for
  this option).
  + Added support to subdivide selections into several pages in order to
  exactly fit the screen size of specific devices (with eReaders, such as my
  Nook, in mind).


krop (0.1.0)  2010-08-19 -- Armin Straub

  * Initial personal release.



krop/krop.py
#! /usr/bin/env python
# -*- coding: iso-8859-1 -*-

"""
krop: A tool to crop PDF files

Copyright (C) 2010-2016 Armin Straub, http://arminstraub.com
"""

"""
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.
"""

import sys

from version import __version__
from config import KDE


def main():
    from argparse import ArgumentParser, RawTextHelpFormatter
    parser = ArgumentParser(description=__doc__,
            formatter_class=RawTextHelpFormatter)

    parser.add_argument('-v', '--version', action='version', version='%(prog)s ' + __version__)

    parser.add_argument('file', nargs='?', help='PDF file to open')
    parser.add_argument('-o', '--output', help='where to save the cropped PDF')
    parser.add_argument('--rotate', type=int, choices=[0,90,180,270], help='how much to rotate the cropped pdf clockwise (default: 0)')
    parser.add_argument('--whichpages', help='which pages (e.g. "1-5" or "1,3-") to include in cropped PDF (default: all)')
    parser.add_argument('--initialpage', help='which page to open initially (default: 1)')
    parser.add_argument('--autotrim', action='store_true', help='create a selection for the entire initial page minus blank margins')
    parser.add_argument('--selections', type=str, choices=['all','evenodd','individual'], help='to which pages should selections apply')
    parser.add_argument('--no-kde', action='store_true', help='do not use KDE libraries (default: use if available)')
    parser.add_argument('--no-PyPDF2', action='store_true', help='do not use PyPDF2 instead of pyPdf (default: use PyPDF2 if available)')

    args = parser.parse_args()

    # start the GUI
    if KDE:
        from PyKDE4.kdecore import ki18n, KCmdLineArgs, KAboutData
        from PyKDE4.kdeui import KApplication
        appName     = "krop"
        catalog     = ""
        programName = ki18n("krop")
         
        aboutData = KAboutData(appName, catalog, programName, __version__)
         
        KCmdLineArgs.init(aboutData)
        app = KApplication()
    else:
        from PyQt4.QtGui import QApplication
        app = QApplication(sys.argv)
        app.setApplicationName("krop")

    app.setOrganizationName("arminstraub.com")
    app.setOrganizationDomain("arminstraub.com")

    from mainwindow import MainWindow
    window=MainWindow()

    if args.file is not None:
        fileName = args.file.decode(sys.stdin.encoding or sys.getdefaultencoding())
        window.openFile(fileName)
    if args.output is not None:
        window.ui.editFile.setText(args.output)
    if args.whichpages is not None:
        window.ui.editWhichPages.setText(args.whichpages)
    if args.rotate is not None:
        window.ui.comboRotation.setCurrentIndex({0:0,90:2,180:3,270:1}[args.rotate])
    if args.selections is not None:
        if args.selections == 'all':
            window.ui.radioSelAll.setChecked(True)
        elif args.selections == 'evenodd':
            window.ui.radioSelEvenOdd.setChecked(True)
        elif args.selections == 'individual':
            window.ui.radioSelIndividual.setChecked(True)
    if args.initialpage is not None:
        window.ui.editCurrentPage.setText(args.initialpage)
        window.slotCurrentPageEdited(args.initialpage)
    if args.autotrim:
        window.slotTrimMarginsAll()

    # shut down on ctrl+c when pressed in terminal (not gracefully, though)
    # http://stackoverflow.com/questions/4938723/
    import signal
    signal.signal(signal.SIGINT, signal.SIG_DFL)

    window.show()
    sys.exit(app.exec_())

if __name__ == "__main__":
    main()
